### PR TITLE
ODIS monitoring improvements + Komenci support

### PR DESCRIPTION
**Pull request by alecps**
### Description

Komenci Changes: Update the process by which ODIS signers calculate quota to include balances and txCounts of walletAddresses. Before quota is calculated the celoBalance, cusdBalance and txCount inputs are summed across the accountAddress and walletAddress. This change is necessary so that users who are onboarded with komenci can query odis even though their balance is held in a walletAddress. 

Monitoring Changes: Adds a monitoring bot that regularly queries the ODIS combiner to confirm that the system is functioning properly. Adds structured logging across the entire phone-number-privacy package. Updates Azure templates to use a user defined identity to fix issue where key retrieval was failing on restart. 

### Other changes

Refactors various logic into the common package 

### Tested

On alfajoresstaging

### Related issues

NA

### Backwards compatibility

Yes

### CODE REVIEW

File: packages/phone-number-privacy/combiner/src/signing/get-threshold-signature.ts [hunk 1]
```diff
@@ -225,7 +231,7 @@ function requestSignature(
 
 function getMajorityErrorCode(errorCodes: Map<number, number>) {
   if (errorCodes.size > 1) {
-    logger.error(ErrorMessage.INCONSISTENT_SIGNER_RESPONSES)
+    logger.error({ errorCodes }, ErrorMessage.INCONSISTENT_SIGNER_RESPONSES)
```
Comment (codyborn) [inline packages/phone-number-privacy/combiner/src/signing/get-threshold-signature.ts:234]: Based on [Bunyan docs](https://github.com/trentm/node-bunyan#log-method-api), it sounds like this should be a named parameter.
```suggestion
    logger.error({ errorCodes: errorCodes }, ErrorMessage.INCONSISTENT_SIGNER_RESPONSES)
```

File: packages/phone-number-privacy/combiner/src/signing/get-threshold-signature.ts [hunk 2]
```diff
@@ -180,7 +185,7 @@ function logResponseDiscrepancies(responses: SignMsgRespWithStatus[]) {
           totalQuota: response.signMessageResponse.totalQuota,
         }
       })
-      logger.error(`Discrepancy found in signers' measured quota values ${values}`)
+      logger.error({ values }, `Discrepancy found in signers' measured quota values`)
```
Comment (codyborn) [inline packages/phone-number-privacy/combiner/src/signing/get-threshold-signature.ts:188]: I think this should be named as well.

File: packages/phone-number-privacy/common/src/utils/authentication.ts [hunk 3]
```diff
@@ -24,7 +24,7 @@ export async function authenticateUser(
   // https://tools.ietf.org/html/rfc7235#section-4.2
   const messageSignature = request.get('Authorization')
   const message = JSON.stringify(request.body)
-  const signer = request.body.account
+  const signer = JSON.stringify(request.body.account)
```
Comment (codyborn) [inline packages/phone-number-privacy/common/src/utils/authentication.ts]: Why was this necessary? Just wondering because in the case that this wasn't working before, we shouldn't take a dependency on this published version of pgpnp-common for this method.

File: packages/phone-number-privacy/common/src/utils/config-utils.ts [hunk 4]
```diff
@@ -1,5 +1,21 @@
...
+
+export function fetchEnv(name: string): string {
+  if (process.env[name] === undefined || process.env[name] === '') {
+    throw new Error(`ENV var '${name}' was not defined`)
+  }
+  return process.env[name] as string
+}
+
+export function fetchEnvOrDefault(name: string, defaultValue: string): string {
+  return process.env[name] === undefined || process.env[name] === ''
```
Comment (codyborn) [inline packages/phone-number-privacy/common/src/utils/config-utils.ts:18]:
```suggestion
  return process.env[name] || process.env[name] === ''
```

File: packages/attestation-service/src/metrics.ts [hunk 5]
```diff
@@ -1,72 +0,0 @@
-import { Counter, Gauge } from 'prom-client'
```
Comment (codyborn) [inline packages/attestation-service/src/metrics.ts]: Was this file intentionally removed?

File: packages/phone-number-privacy/monitor/src/query.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+contractKit.addAccount(privateKey)
+contractKit.defaultAccount = accountAddress
+
+export const queryOdisForSalt = () => {
+  const authSigner: AuthSigner = {
+    authenticationMethod: OdisUtils.Query.AuthenticationMethod.WALLET_KEY,
+    contractKit,
+  }
+
+  return OdisUtils.PhoneNumberIdentifier.getPhoneNumberIdentifier(
```
Comment (codyborn) [inline packages/phone-number-privacy/monitor/src/query.ts:21]: Can you pass in a client version string with something that indicates this is coming from a canary test and not the mobile client (ex. "canary:1.0.0")? This way we can tell from the Combiner logs that the out-of-quotas are not coming from real users.

File: packages/phone-number-privacy/monitor/package.json [hunk 7]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "@celo/phone-number-privacy-monitor",
+  "version": "1.22.5",
```
Comment (codyborn) [inline packages/phone-number-privacy/monitor/package.json]: Since this is new, we can start it on 1.0.0.

File: packages/phone-number-privacy/signer/src/database/wrappers/account.ts [hunk 8]
```diff
@@ -76,7 +75,7 @@ export async function getDidMatchmaking(account: string): Promise<boolean> {
  * Set did matchmaking to true in database.  If record doesn't exist, create one.
  */
 export async function setDidMatchmaking(account: string) {
-  logger.debug('Setting did matchmaking')
+  logger.debug({ account }, 'Setting did matchmaking')
```
Comment (codyborn) [inline packages/phone-number-privacy/signer/src/database/wrappers/account.ts:81]: Should be a named param?

File: packages/phone-number-privacy/signer/src/database/wrappers/request.ts [hunk 9]
```diff
@@ -12,9 +11,10 @@ export async function getRequestExists(
   request: GetBlindedMessagePartialSigRequest
 ): Promise<boolean> {
   if (!request.timestamp) {
+    logger.debug('request does not have timestamp')
     return false // TODO(Alec) make timestamps required
   }
-  logger.debug('Checking if request exists')
+  logger.debug({ request }, 'Checking if request exists')
```
Comment (codyborn) [inline packages/phone-number-privacy/signer/src/database/wrappers/request.ts:17]: Same here

File: packages/phone-number-privacy/common/src/utils/logger.ts [hunk 10]
```diff
@@ -1,12 +1,27 @@
-// tslint:disable: no-console
+import Logger, { createLogger, levelFromName, LogLevelString, stdSerializers } from 'bunyan'
+import bunyanDebugStream from 'bunyan-debug-stream'
+import { createStream } from 'bunyan-gke-stackdriver'
```
Comment (codyborn) [inline packages/phone-number-privacy/common/src/utils/logger.ts:3]: Did you see who created this package?
https://www.npmjs.com/package/bunyan-gke-stackdriver
Thanks @mcortesi 😄

File: packages/phone-number-privacy/signer/.env [hunk 11]
```diff
@@ -17,4 +17,13 @@ DB_USE_SSL=true
 KEYSTORE_TYPE=MockSecretManager
 KEYSTORE_GOOGLE_PROJECT_ID=mockProjectId
 KEYSTORE_GOOGLE_SECRET_NAME=mockSecretName
-KEYSTORE_GOOGLE_SECRET_VERSION=latest
\ No newline at end of file
+KEYSTORE_GOOGLE_SECRET_VERSION=latest
+# Options: json, human (default), stackdriver
+LOG_FORMAT=stackdriver
```
Comment (codyborn) [inline packages/phone-number-privacy/signer/.env:22]: We run our signers in Azure Container Instance and pipe the logs to [Azure Log Analytics](https://portal.azure.com#@7cb7628a-e37c-4afb-8332-2029e418980e/blade/Microsoft_Azure_Monitoring_Logs/LogsBlade/resourceId/%2Fsubscriptions%2F97e2b592-255b-4f92-bce0-127257163c36%2Fresourcegroups%2Fpgpnp-alfajores%2Fproviders%2Fmicrosoft.operationalinsights%2Fworkspaces%2Fpgpnp-alfajores/source/LogsBlade.AnalyticsShareLinkToQuery/q/H4sIAAAAAAAAA3POzytJzMxLLfLMKy5JzEtO9clPj3f2AQCR58T%252FFwAAAA%253D%253D/timespan/P1D). We should double check that this formatting is compatible with [our alerting](https://portal.azure.com/#@mcelo.onmicrosoft.com/resource/subscriptions/97e2b592-255b-4f92-bce0-127257163c36/resourceGroups/pgpnp-alfajores/providers/Microsoft.OperationalInsights/workspaces/pgpnp-alfajores/alerts).

File: packages/phone-number-privacy/signer/src/index.ts [hunk 12]
```diff
@@ -18,6 +18,6 @@ async function start() {
 }
 
 start().catch((e) => {
-  logger.error('Fatal error occured. Exiting', e)
+  logger.error({ e }, 'Fatal error occured. Exiting')
```
Comment (codyborn) [inline packages/phone-number-privacy/signer/src/index.ts]:
```suggestion
  logger.error({ err: e }, 'Fatal error occured. Exiting')
```

File: packages/phone-number-privacy/combiner/src/index.ts [hunk 13]
```diff
@@ -1,31 +1,36 @@
...
 
+require('dotenv').config()
+
 // DEPRECATED, TODO: Remove once clients are all on contract kit version 4.11
 export const getDistributedBlindedSalt = functions
   .region('us-central1', 'europe-west3')
-  .https.onRequest(async (request, response) => {
-    logger.info('Begin getDistributedBlindedSalt request')
-    return handleGetBlindedMessageSig(request, response)
+  .https.onRequest(async (req, res) => {
```
Comment (codyborn) [inline packages/phone-number-privacy/combiner/src/index.ts:12]: In general, it might be useful to include a sessionId/correlationId in each logging event so that we can connect the events together after the fact. We don't have to do it in this PR but just something to think about.

**Review by codyborn — CHANGES_REQUESTED** [review]
Nice work! Added some small feedback.

File: packages/phone-number-privacy/combiner/src/signing/get-threshold-signature.ts (hunk 1 above)
Comment (alecps) [reply packages/phone-number-privacy/combiner/src/signing/get-threshold-signature.ts:234]: I *think* that the field will assume the name of the variable based on how this is done in the attestation-service https://github.com/celo-org/celo-monorepo/blob/e1a6a102dd1654a6774d47198d29e8ccb2360887/packages/attestation-service/src/db.ts#L105 I'll make sure to check.

File: packages/attestation-service/src/metrics.ts (hunk 5 above)
Comment (alecps) [reply packages/attestation-service/src/metrics.ts]: Nope! Good catch 😅

File: packages/phone-number-privacy/common/src/utils/logger.ts (hunk 10 above)
Comment (alecps) [reply packages/phone-number-privacy/common/src/utils/logger.ts:3]: 😂 🙏 a true hero

File: packages/phone-number-privacy/combiner/src/index.ts (hunk 13 above)
Comment (alecps) [reply packages/phone-number-privacy/combiner/src/index.ts:12]: Good idea. Are you thinking something like each request has an id so that we can link all the logs originating from that request together?

File: packages/phone-number-privacy/common/src/utils/config-utils.ts (hunk 4 above)
Comment (alecps) [reply packages/phone-number-privacy/common/src/utils/config-utils.ts:18]: Did you mean to put '!' in front of process.env? I think we might want to be careful in case of env variables with the value 'false' or '0'. I *think* js always treats these as strings so probably fine.

File: packages/phone-number-privacy/common/src/utils/authentication.ts (hunk 3 above)
Comment (alecps) [reply packages/phone-number-privacy/common/src/utils/authentication.ts]: It probably wasn't necessary come to think of it. I was just trying to make sure that we didn't try and give bunyan something other than a string there but I don't think it matters

File: packages/phone-number-privacy/signer/.env (hunk 11 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/.env:22]: Good idea. I'll test on staging and use the 'json' format if not

File: packages/phone-number-privacy/monitor/src/query.ts (hunk 6 above)
Comment (alecps) [reply packages/phone-number-privacy/monitor/src/query.ts:21]: Good idea

File: .env [hunk 14]
```diff
@@ -175,8 +175,3 @@ VOTING_BOT_SCORE_SENSITIVITY=5
 # is not easily represented.
 # This schedule is "every 15 minutes", which is close enough. For some epochs, it will run twice
 VOTING_BOT_CRON_SCHEDULE="*/15 * * * *"
-
-# ODIS Service Tests
-ODIS_SIGNER_SERVICE_URL=https://staging-pgpnp-signer0.azurefd.net
```
Comment (codyborn) [inline .env:180]: We should keep these here so that Vscode debugging works. We can have a copy in both locations if needed.

File: packages/phone-number-privacy/combiner/src/signing/get-threshold-signature.ts (hunk 1 above)
Comment (alecps) [reply packages/phone-number-privacy/combiner/src/signing/get-threshold-signature.ts:234]: I've verified that the variable name becomes the name of the parameter

File: .env (hunk 14 above)
Comment (alecps) [reply .env:180]: 👍

File: packages/phone-number-privacy/signer/azure-templates/container-parameters.json [hunk 15]
```diff
@@ -90,6 +98,9 @@
         },
         "logAnalyticsWorkspaceKey": {
             "value": "TODO"
-        }
+        },
+        "userAssignedIdentity": {
```
Comment (codyborn) [inline packages/phone-number-privacy/signer/azure-templates/container-parameters.json:102]: Thanks!! Can you also update the `azure-templates/README.md` with the prerequisite step to deploy the UserAssignedIdentity?

Comment (alecps) [reply packages/phone-number-privacy/signer/azure-templates/container-parameters.json:102]: done!

File: packages/phone-number-privacy/signer/src/signing/query-quota.ts [hunk 16]
```diff
@@ -80,25 +90,41 @@ export async function getRemainingQueryCount(
  * If the caller is not verified, they must have a minimum balance to get the unverifiedQueryMax.
  */
 async function getQueryQuota(account: string, hashedPhoneNumber?: string) {
-  if (hashedPhoneNumber && (await isVerified(account, hashedPhoneNumber))) {
-    logger.debug('Account is verified')
-    const transactionCount = await getTransactionCountFromAccount(account)
-    return (
+  let walletAddress = await getWalletAddress(account)
+  logger.debug({ account, walletAddress }, 'begin getQueryQuota')
+  if (account === walletAddress) {
```
Comment (codyborn) [inline packages/phone-number-privacy/signer/src/signing/query-quota.ts]: It's probably safest to do a normalized comparison
```suggestion
  if (account.toLowerCase() === walletAddress.toLowerCase()) {
```

**Review by codyborn — APPROVED** [review]
Looks great!
